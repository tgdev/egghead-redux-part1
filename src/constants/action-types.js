export const ACTION_ADD_TODO = 'ADD_TODO';
