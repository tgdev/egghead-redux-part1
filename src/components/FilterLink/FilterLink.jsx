import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { ACTION_SET_VISIBILITY_FILTER } from 'constants/action-types';

import store from '../../store';

class FilterLink extends Component {
  constructor() {
    super();
    this.handleSetFilter = this.handleSetFilter.bind(this);
  }

  handleSetFilter() {
    store.dispatch({
      type: ACTION_SET_VISIBILITY_FILTER,
      filter: this.props.filter
    });
  }

  render() {
    const { filter, children, currentFilter } = this.props;

    if (filter === currentFilter) {
      return <span>{children}</span>;
    }

    return (
      <button onClick={this.handleSetFilter}>
        {children}
      </button>
    );
  }
}

FilterLink.propTypes = {
  filter: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired,
  currentFilter: PropTypes.string.isRequired
};

export default FilterLink;
