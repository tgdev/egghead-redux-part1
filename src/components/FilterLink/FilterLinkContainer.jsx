import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { ACTION_SET_VISIBILITY_FILTER } from 'constants/action-types';

import FilterLink from './FilterLink';

class FilterLinkContainer extends Component {
  componentDidMount() {
    const { store } = this.context;
    store.subscribe(() => {
      this.forceUpdate();
    });
  }

  componentWillUnmount() {
    this.unsubscribe();
  }

  render() {
    const { store } = this.context;
    const { filter, children } = this.props;
    const { visibilityFilter } = store.getState();
    const isActive = filter === visibilityFilter;

    return (
      <FilterLink
        active={isActive}
        handleFilter={() => {
          store.dispatch({
            type: ACTION_SET_VISIBILITY_FILTER,
            filter
          });
        }}>
        {children}
      </FilterLink>
    );
  }
}

FilterLinkContainer.propTypes = {
  filter: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired
};

FilterLinkContainer.contextTypes = {
  store: PropTypes.object
};

export default FilterLinkContainer;
