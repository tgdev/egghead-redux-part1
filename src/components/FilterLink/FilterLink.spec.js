import React from 'react';
import { shallow } from 'enzyme';

import { FILTER_OPTION_ALL } from 'constants/visibility-filter-options';

import FilterLink from './FilterLink';

describe('FilterLink component', () => {

  const mockFilterOption = FILTER_OPTION_ALL;

  it('renders FilterLink correctly', () => {
    const wrapper = shallow(<FilterLink filter={mockFilterOption}>All</FilterLink>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders FilterLink with children', () => {
    const wrapper = shallow(<FilterLink filter={mockFilterOption}>All</FilterLink>);
    expect(wrapper.children().length).toBe(1);
  });

  // it('renders FilterLink with filter prop', () => {
  //   const wrapper = shallow(<FilterLink filter={mockFilterOption}>All</FilterLink>);
  //   const filterProp = wrapper.instance().props.filter;
  //   expect(filterProp).toEqual(FILTER_OPTION_ALL);
  // });

  // it('triggers a function when clicked', () => {
  //   const wrapper = shallow(<FilterLink filter={mockFilterOption}>All</FilterLink>);
  //   const link = wrapper.find('a');
  //   const mockEvent = { preventDefault: jest.fn() };
  //   link.simulate('click', mockEvent);
  //   expect(link).toHaveBeenCalledTimes(1);
  // });

});
