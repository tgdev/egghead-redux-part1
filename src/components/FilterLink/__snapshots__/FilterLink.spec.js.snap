// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterLink component renders FilterLink correctly 1`] = `
<button
  onClick={[Function]}
>
  All
</button>
`;
